core/src/main/java/io/grpc/Metadata.java

```diff
@@ -132,87 +135,145 @@ public Integer parseAsciiString(String serialized) {
     }
   };
 
-  /** All value lists can be added to. No value list may be empty. */
-  // Use LinkedHashMap for consistent ordering for tests.
-  private final Map<String, List<MetadataEntry>> store =
-      new LinkedHashMap<String, List<MetadataEntry>>();
-
-  /** The number of headers stored by this metadata.  */
-  private int storeCount;
+  /**
+   * Constructor called by the transport layer when it receives binary metadata.
+   */
+  Metadata(byte[]... binaryValues) {
+    this(binaryValues.length / 2, binaryValues);
+  }
 
   /**
    * Constructor called by the transport layer when it receives binary metadata.
    */
   // TODO(louiscryan): Convert to use ByteString so we can cache transformations
-  @Internal
-  public Metadata(byte[]... binaryValues) {
-    checkArgument(binaryValues.length % 2 == 0,
-        "Odd number of key-value pairs: %s", binaryValues.length);
-    for (int i = 0; i < binaryValues.length; i += 2) {
-      // The transport might provide an array with null values at the end.
-      if (binaryValues[i] == null) {
-        break;
-      }
-      String name = new String(binaryValues[i], US_ASCII);
-      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[i + 1]));
-    }
+  Metadata(int usedNames, byte[]... binaryValues) {
+    // TODO(carl-mastrangelo): maybe remove this, unless
+    // https://bugs.openjdk.java.net/browse/JDK-6445664 is fixed.
```

Comment:
Out of curiosity I did some digging. In `byteCodeInfo.cpp` there is a method `bool InlineTree::try_to_inline(...)` which has the check
```cpp
// don't inline into giant methods
if (C->over_inlining_cutoff()) {
  return false;
}
```

and in `compile.cpp`

```cpp
  bool over_inlining_cutoff() const {
    if (!inlining_incrementally()) {
      return unique() > (uint)NodeCountInliningCutoff;
    } else {
      return live_nodes() > (uint)LiveNodeCountInliningCutoff;
    }
  }
```

Then there is the JVM option `-XX:LiveNodeCountInliningCutoff `, which this method seems to enforce. On my machine that  option is set to 40000.

Assuming I got the right code, it seems that it's not bytecodes that count when determining if a caller is too big to have callees inlined, but the method representation by the compiler.